src/dos/cdrom.cpp

```diff
@@ -261,3 +262,212 @@ void CDROM_Interface::LagDriveResponse() const
 
 	prev_ticks = PIC_Ticks;
 }
+
+void CDROM_Interface_Physical::CdReaderLoop()
+{
+	std::unique_lock lock(mutex);
+	while (true) {
+		waiter.wait(lock, [&]{ return sectors_remaining > 0 || should_exit; });
+
+		if (should_exit) {
+			return;
+		}
+
+		if (!queue.IsRunning()) {
+			queue.Clear();
+			queue.Start();
+		}
+
+		// Copy these to local variables so that we can release the mutex lock.
+		const uint32_t local_sectors_remaining = sectors_remaining;
+		const uint32_t local_sector_start = current_sector;
+		lock.unlock();
+
+		// Slow call to read from the CDROM drive. Avoid holding any locks
+		const std::vector<int16_t> audio_samples = ReadAudio(local_sector_start, local_sectors_remaining);
+		const size_t num_frames = audio_samples.size() / 2;
+		const size_t sectors_read = audio_samples.size() / SAMPLES_PER_REDBOOK_FRAME;
+
+		// Convert from int16_t to float before enqueuing
+		std::vector<AudioFrame> audio_frames;
+		for (size_t i = 0; i < num_frames; ++i) {
+			const int16_t left = host_to_le16(audio_samples[i * 2]);
+			const int16_t right = host_to_le16(audio_samples[(i * 2) + 1]);
+			audio_frames.emplace_back(left, right);
+		}
+
+		queue.BulkEnqueue(audio_frames, audio_frames.size());
+
+		lock.lock();
+		if (mixer_channel && !is_paused && sectors_remaining > 0 && queue.IsRunning()) {
+			mixer_channel->Enable(true);
+		}
+		if (sectors_remaining == local_sectors_remaining && current_sector == local_sector_start) {
+			sectors_remaining -= sectors_read;
+			current_sector += sectors_read;
+		}
+	}
+}
+
+void CDROM_Interface_Physical::InitAudio()
+{
+	if (mixer_channel) {
+		return;
+	}
+
+	bool found_unique_name = false;
+	std::string name;
+	for (int i = 0; i < 32; ++i) {
```

Comment:
Could use the `MaxCdAudioChannels` constant here I asked for a couple of commits earlier 😄

Reply:
This is the same code moved from the Linux class to the platform independent class.  I added that constant the other day.